File: c/bulk.cpp
#include <string>
#include <tr1/unordered_set>
#include <map>
#include <vector>
#include <queue>
#include <iostream>
#include <fstream>
#include <sys/time.h>

using namespace std;
using namespace tr1;

unsigned int getMicroSecs()
{
	struct timeval timeVal;
	gettimeofday(&timeVal, 0);
	return timeVal.tv_sec * 1000000 + timeVal.tv_usec;
}

enum Orientation
{
	xFace,
	yFace,
	zFace
};

enum Direction
{
	leftDir,
	rightDir,
	upDir,
	downDir,
	frontDir,
	backDir
};

// ---------------------------------------------------------------------

typedef unsigned long long Vertex;

typedef unordered_set<Vertex> VectorSet; 

short& x(Vertex& v)
{
	short* x = (short*) &v;
	return *x;
}

short& y(Vertex& v)
{
	short* y = (short*) &v;
	y+=1;
	return *y;
}

short& z(Vertex& v)
{
	short* z = (short*) &v;
	z+=2;
	return *z;
}

Vertex vertex(short xV, short yV, short zV)
{
	Vertex v = 0;
	x(v) = xV;
	y(v) = yV;
	z(v) = zV;
	return v;
}

// ---------------------------------------------------------------------

class Face
{
public:
	VectorSet squares;
	vector<Vertex> cornerVertices;
	Orientation orientation;

public:
	void deriveSquares();

private:
	void determineOrientation();
	void collectSquares(Vertex start);
	void tryGoUp(queue<Vertex> &queue, Vertex &v);
	void tryGoDown(queue<Vertex> &queue, Vertex &v);
	void tryGoLeft(queue<Vertex> &queue, Vertex &v);
	void tryGoRight(queue<Vertex> &queue, Vertex &v);

	bool crossesBorder(Vertex from, Vertex to, Direction dir);
};

void Face::determineOrientation()
{
	bool xChanges = false;
	bool yChanges = false;
	bool zChanges = false;
	Vertex v = cornerVertices.front();
	for (vector<Vertex>::iterator iter = cornerVertices.begin(); iter != cornerVertices.end(); iter++)
	{
		xChanges |= x(v) != x(*iter);
		yChanges |= y(v) != y(*iter);
		zChanges |= z(v) != z(*iter);
	}

	if(xChanges && yChanges)
		orientation = zFace;
	if(xChanges && zChanges)
		orientation = yFace;
	if(yChanges && zChanges)
		orientation = xFace;
}

void Face::deriveSquares()
{
	squares.clear();

	determineOrientation();

	vector<Vertex>::iterator iter = cornerVertices.begin();
	Vertex min = *iter;
	while(iter != cornerVertices.end())
	{
		if(*iter < min)
			min = *iter;
		iter++;
	}

	// min = starting point
	squares.insert(min);
	collectSquares(min);
}

void Face::collectSquares(Vertex start)
{
	// try all 4 directions
	queue<Vertex> queue;
	queue.push(start);

	while(!queue.empty())
	{
		Vertex v = queue.front();
		tryGoLeft(queue, v);
		tryGoRight(queue, v);
		tryGoUp(queue, v);
		tryGoDown(queue, v);
		queue.pop();
	}
}

void Face::tryGoLeft(queue<Vertex> &queue, Vertex &v)
{
	Vertex next = v;
	switch(orientation)
	{
		case zFace:
		case yFace:
			x(next)--;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, leftDir))
				return;
			break;
		case xFace:
			y(next)--;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, downDir))
			    return;
			break;
	}

	squares.insert(next);
	queue.push(next);
}

void Face::tryGoRight(queue<Vertex> &queue, Vertex &v)
{
	Vertex next = v;
	switch(orientation)
	{
		case zFace:
		case yFace:
			x(next)++;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, rightDir))
				return;
			break;
		case xFace:
			y(next)++;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, upDir))
				return;
			break;
	}

	squares.insert(next);
	queue.push(next);
}

void Face::tryGoDown(queue<Vertex> &queue, Vertex &v)
{
	Vertex next = v;
	switch(orientation)
	{
		case zFace:
			y(next)--;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, downDir))
				return;
			break;
		case xFace:
		case yFace:
			z(next)--;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, backDir))
				return;
			break;
	}

	squares.insert(next);
	queue.push(next);
}

void Face::tryGoUp(queue<Vertex> &queue, Vertex &v)
{
	Vertex next = v;
	switch(orientation)
	{
		case zFace:
			y(next)++;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, upDir))
				return;
			break;
		case xFace:
		case yFace:
			z(next)++;
			if(squares.count(next))
				return;
			if(crossesBorder(v, next, frontDir))
				return;
			break;
	}

	squares.insert(next);
	queue.push(next);
}

bool Face::crossesBorder(Vertex from, Vertex to, Direction dir)
{
	Vertex v1 = cornerVertices.back();
	for(vector<Vertex>::iterator iter = cornerVertices.begin(); iter != cornerVertices.end(); iter++)
	{
		Vertex v2 = *iter;
		Vertex v = to;
		switch(dir)
		{
			case leftDir:
				v = from;
			case rightDir:
				if( x(v1) != x(v2) || x(v1) != x(v))
					break;
				switch(orientation)
				{
					case zFace:
						if(y(v1) <= y(v) && y(v) < y(v2) || y(v2) <= y(v) && y(v) < y(v1))
						    return true;
						break;
					case yFace:
						if(z(v1) <= z(v) && z(v) < z(v2) || z(v2) <= z(v) && z(v) < z(v1))
						    return true;
						break;
				}
				break;
			case downDir:
				v = from;
			case upDir:
				if(y(v1) != y(v2) || y(v1) != y(v))
					break;
				switch(orientation)
				{
					case zFace:
						if(x(v1) <= x(v) && x(v) < x(v2) || x(v2) <= x(v) && x(v) < x(v1))
						    return true;
						break;
					case xFace:
						if(z(v1) <= z(v) && z(v) < z(v2) || z(v2) <= z(v) && z(v) < z(v1))
						    return true;
						break;
				}
				break;
			case backDir:
				v = from;
			case frontDir:
				if(z(v1) != z(v2) || z(v1) != z(v))
					break;
				switch(orientation)
				{
					case yFace:
						if(x(v1) <= x(v) && x(v) < x(v2) || x(v2) <= x(v) && x(v) < x(v1))
						    return true;
						break;
					case xFace:
						if(y(v1) <= y(v) && y(v) < y(v2) || y(v2) <= y(v) && y(v) < y(v1))
						    return true;
						break;
				}
				break;
		}
		v1 = v2;
	}
	return false;
}

// ---------------------------------------------------------------------

vector<Face*> faces;
VectorSet blocks;
map<Orientation , VectorSet*> borderSquares;

void tryGoBack(queue<Vertex> &queue, Vertex v)
{
	if(borderSquares[zFace]->count(v))
		return;

	z(v)--;

	if(blocks.count(v))
		return;

	blocks.insert(v);
	queue.push(v);
}

void tryGoFront(queue<Vertex> &queue, Vertex v)
{
	z(v)++;

	if(borderSquares[zFace]->count(v))
		return;

	if(blocks.count(v))
		return;

	blocks.insert(v);
	queue.push(v);
}

void tryGoLeft(queue<Vertex> &queue, Vertex v)
{
	if(borderSquares[xFace]->count(v))
		return;

	x(v)--;

	if(blocks.count(v))
		return;

	blocks.insert(v);
	queue.push(v);
}

void tryGoRight(queue<Vertex> &queue, Vertex v)
{
	x(v)++;

	if(borderSquares[xFace]->count(v))
		return;

	if(blocks.count(v))
		return;

	blocks.insert(v);
	queue.push(v);
}

void tryGoDown(queue<Vertex> &queue, Vertex v)
{
	if(borderSquares[yFace]->count(v))
		return;

	y(v)--;

	if(blocks.count(v))
		return;

	blocks.insert(v);
	queue.push(v);
}

void tryGoUp(queue<Vertex> &queue, Vertex v)
{
	y(v)++;

	if(borderSquares[yFace]->count(v))
		return;

	if(blocks.count(v))
		return;

	blocks.insert(v);
	queue.push(v);
}

void collectBlocks(Vertex start)
{
	queue<Vertex> queue;
	blocks.insert(start);
	queue.push(start);

	while(!queue.empty())
	{
		Vertex v = queue.front();
		tryGoLeft(queue, v);
		tryGoRight(queue, v);
		tryGoDown(queue, v);
		tryGoUp(queue, v);
		tryGoBack(queue, v);
		tryGoFront(queue, v);
		queue.pop();
	}
}

void solve()
{
	borderSquares[xFace]->clear();
	borderSquares[yFace]->clear();
	borderSquares[zFace]->clear();

	for (vector<Face*>::iterator iter = faces.begin(); iter != faces.end(); iter++)
	{
		// derive all Squares of this face
		Face* f = *iter;
//		unsigned int startTime = getMicroSecs();
//		cout << "deriveSquares: ";
		f->deriveSquares();
//		cout << (getMicroSecs() - startTime) << endl;

		// put all squares of faces with the same orientation together
		VectorSet* squares = borderSquares[f->orientation];
		for(VectorSet::iterator iter = f->squares.begin(); iter != f->squares.end(); iter++)
		{
			Vertex v = *iter;
			squares->insert(v);
		}
	}

	VectorSet::iterator iter = borderSquares[zFace]->begin();
	Vertex min = *iter;
	while(iter != borderSquares[zFace]->end())
	{
		if(*iter < min)
			min = *iter;
		iter++;
	}

	blocks.clear();
	blocks.insert(min);
//	unsigned int startTime = getMicroSecs();
//		cout << "collectBlocks: ";
		collectBlocks(min);
//		cout << (getMicroSecs() - startTime) << endl;

	
	cout << "The bulk is composed of " << blocks.size() << " units.\n";
}

int main(int argc, char** args)
{
	borderSquares[xFace] = new VectorSet();
	borderSquares[yFace] = new VectorSet();
	borderSquares[zFace] = new VectorSet();

	int numTests;
	istream* input = &cin;
	if(argc > 1)
		input = new ifstream("/work/workspace/SphereOnlineJudge/testcases/BULK.txt");
	(*input) >> numTests;
	while(numTests-->0)
	{
		faces.clear();
		int numFaces;
		(*input) >> numFaces;
		while(numFaces-->0)
		{
			int numVertices;
			(*input) >> numVertices;
			Face *face = new Face();
			while(numVertices-->0)
			{
				short xV,yV,zV;
				(*input) >> xV >> yV >> zV;
				face->cornerVertices.push_back(vertex(xV, yV, zV));
			}
			faces.push_back(face);
		}
		solve();

		for(vector<Face*>::iterator iter = faces.begin(); iter != faces.end(); iter++)
			delete *iter;
	}
}

File: java/Bulk.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;

public class Bulk 
{
	private static final HashSet<Face> faces = new HashSet<Face>();
	private static final HashSet<Vertex> blocks = new HashSet<Vertex>();
	private static final HashMap<Orientation, HashSet<Vertex>> borderSquares = new HashMap<Orientation, HashSet<Vertex>>();
	
	public static void main(String[] args) throws IOException 
	{
		borderSquares.put(Orientation.xyFace, new HashSet<Vertex>());
		borderSquares.put(Orientation.xzFace, new HashSet<Vertex>());
		borderSquares.put(Orientation.yzFace, new HashSet<Vertex>());
		
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		if(!reader.ready())
			reader = new BufferedReader(new FileReader("testcases/BULK.txt"));
		
		int numTests = Integer.parseInt(reader.readLine());
		while(numTests-->0)
		{
			faces.clear();
			int numFaces = Integer.parseInt(reader.readLine());
			while(numFaces-->0)
			{
				String line = reader.readLine();
				int start = line.indexOf(' ') + 2;
				int end = 0;
				Face face = new Face();
				while(end >= 0)
				{
					end = line.indexOf(' ', start);
					int x = Integer.parseInt(line.substring(start, end));
					start = end + 1;
					end = line.indexOf(' ', start);
					int y = Integer.parseInt(line.substring(start, end));
					start = end + 1;
					end = line.indexOf(' ', start);
					int z = end >= 0 ? Integer.parseInt(line.substring(start, end)) : Integer.parseInt(line.substring(start));
					start = end + 2;
					face.addCornerVertex(x, y, z);
				}
				faces.add(face);
			}
			solve();
		}

	}

	private static void solve() 
	{
		borderSquares.get(Orientation.xyFace).clear();
		borderSquares.get(Orientation.xzFace).clear();
		borderSquares.get(Orientation.yzFace).clear();
		
		for(Face face : faces)
		{
			// derive all Squares of this face
			face.deriveSquares();
			
			// put all squares of faces with the same orientation together
			borderSquares.get(face.getOrientation()).addAll(face.getSquares());
		}
		
		Vertex min = null;
		for(Vertex v : borderSquares.get(Orientation.xyFace))
		{
			if(min == null || v.compareTo(min) < 0)
				min = v;
		}
		
		blocks.clear();
		blocks.add(min);
		collectBlocks(min);
		System.out.println("The bulk is composed of " + blocks.size() + " units.");
		
	}

	private static void collectBlocks(Vertex start) 
	{
		LinkedList<Vertex> queue = new LinkedList<Vertex>();
		queue.add(start);
		
		while(!queue.isEmpty())
		{
			Vertex v = queue.removeFirst();
			tryGoLeft(queue, v);
			tryGoRight(queue, v);
			tryGoDown(queue, v);
			tryGoUp(queue, v);
			tryGoBack(queue, v);
			tryGoFront(queue, v);
		}
	}
	
	private static void tryGoLeft(LinkedList<Vertex> queue, Vertex v) 
	{
		if(borderSquares.get(Orientation.yzFace).contains(v))
			return;

		Vertex next = new Vertex(v.x-1, v.y, v.z);
		if(blocks.contains(next))
			return;
		
		blocks.add(next);
		queue.add(next);
	}

	private static void tryGoRight(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = new Vertex(v.x+1, v.y, v.z);
		if(borderSquares.get(Orientation.yzFace).contains(next))
			return;
		
		if(blocks.contains(next))
			return;
		
		blocks.add(next);
		queue.add(next);
	}
	
	private static void tryGoDown(LinkedList<Vertex> queue, Vertex v) 
	{
		if(borderSquares.get(Orientation.xzFace).contains(v))
			return;

		Vertex next = new Vertex(v.x, v.y-1, v.z);
		if(blocks.contains(next))
			return;
		
		blocks.add(next);
		queue.add(next);
	}
	
	private static void tryGoUp(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = new Vertex(v.x, v.y+1, v.z);
		if(borderSquares.get(Orientation.xzFace).contains(next))
			return;
		
		if(blocks.contains(next))
			return;
		
		blocks.add(next);
		queue.add(next);
	}
	
	private static void tryGoBack(LinkedList<Vertex> queue, Vertex v) 
	{
		if(borderSquares.get(Orientation.xyFace).contains(v))
			return;

		Vertex next = new Vertex(v.x, v.y, v.z-1);
		if(blocks.contains(next))
			return;
		
		blocks.add(next);
		queue.add(next);
	}
	
	private static void tryGoFront(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = new Vertex(v.x, v.y, v.z+1);
		if(borderSquares.get(Orientation.xyFace).contains(next))
			return;
		
		if(blocks.contains(next))
			return;
		
		blocks.add(next);
		queue.add(next);
	}
}

enum Orientation
{
	xyFace,
	xzFace,
	yzFace
}

class Face
{
	private final LinkedList<Vertex> cornerVertices = new LinkedList<Vertex>();
	private final LinkedList<Vertex> borderVertices = new LinkedList<Vertex>();
	private final HashSet<Vertex> squares = new HashSet<Vertex>();
	
	public HashSet<Vertex> getSquares() 
	{
		return squares;
	}

	private Orientation orientation = null;
	
	public Orientation getOrientation()
	{
		if(orientation == null)
		{
			boolean xChanges = false;
			boolean yChanges = false;
			boolean zChanges = false;
			Vertex p = cornerVertices.getFirst();
			for(Vertex p2 : cornerVertices)
			{
				xChanges |= p.x != p2.x;
				yChanges |= p.y != p2.y;
				zChanges |= p.z != p2.z;
			}
			
			if(xChanges && yChanges)
				orientation = Orientation.xyFace;
			if(xChanges && zChanges)
				orientation = Orientation.xzFace;
			if(yChanges && zChanges)
				orientation = Orientation.yzFace;
		}
		return orientation;
	}

	public void addCornerVertex(int x, int y, int z) 
	{
		cornerVertices.add(new Vertex(x,y,z));
	}
	
	public void deriveSquares()
	{
		deriveBorderVertices();
		squares.clear();
		
		Vertex min = null;
		for(Vertex v : cornerVertices)
			if(min == null || v.compareTo(min) < 0)
				min = v;

		// min = starting point
		squares.add(min);
		collectSquares(min);
	}
	
	private void collectSquares(Vertex start) 
	{
		// try all 4 directions
		LinkedList<Vertex> queue = new LinkedList<Vertex>();
		queue.add(start);
		
		while(!queue.isEmpty())
		{
			Vertex v = queue.removeFirst();
			tryGoLeft(queue, v);
			tryGoRight(queue, v);
			tryGoUp(queue, v);		
			tryGoDown(queue, v);
		}
	}
	
	private void tryGoLeft(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = null;
		switch(getOrientation())
		{
			case xyFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z), new Vertex(v.x, v.y+1, v.z)))
					return;
				next = new Vertex(v.x-1, v.y, v.z);
				break;
			case xzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z), new Vertex(v.x, v.y, v.z+1)))
					return;
				next = new Vertex(v.x-1, v.y, v.z);
				break;
			case yzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z), new Vertex(v.x, v.y, v.z+1)))
					return;
				next = new Vertex(v.x, v.y-1, v.z);
				break;
		}
		if(squares.contains(next))
			return;
		
		squares.add(next);
		queue.addLast(next);
	}
	
	private void tryGoRight(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = null;
		switch(getOrientation())
		{
			case xyFace:
				if(borderContainsEdge(new Vertex(v.x+1, v.y, v.z), new Vertex(v.x+1, v.y+1, v.z)))
					return;
				next = new Vertex(v.x+1, v.y, v.z);
				break;
			case xzFace:
				if(borderContainsEdge(new Vertex(v.x+1, v.y, v.z), new Vertex(v.x+1, v.y, v.z+1)))
					return;
				next = new Vertex(v.x+1, v.y, v.z);
				break;
			case yzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y+1, v.z), new Vertex(v.x, v.y+1, v.z+1)))
					return;
				next = new Vertex(v.x, v.y+1, v.z);
				break;
		}
		if(squares.contains(next))
			return;
		
		squares.add(next);
		queue.addLast(next);
	}

	private void tryGoUp(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = null;
		switch(getOrientation())
		{
			case xyFace:
				if(borderContainsEdge(new Vertex(v.x, v.y+1, v.z), new Vertex(v.x+1, v.y+1, v.z)))
					return;
				next = new Vertex(v.x, v.y+1, v.z);
				break;
			case xzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z+1), new Vertex(v.x+1, v.y, v.z+1)))
					return;
				next = new Vertex(v.x, v.y, v.z+1);
				break;
			case yzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z+1), new Vertex(v.x, v.y+1, v.z+1)))
					return;
				next = new Vertex(v.x, v.y, v.z+1);
				break;
		}
		if(squares.contains(next))
			return;
		
		squares.add(next);
		queue.addLast(next);
	}
	
	private void tryGoDown(LinkedList<Vertex> queue, Vertex v) 
	{
		Vertex next = null;
		switch(getOrientation())
		{
			case xyFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z), new Vertex(v.x+1, v.y, v.z)))
					return;
				next = new Vertex(v.x, v.y-1, v.z);
				break;
			case xzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z), new Vertex(v.x+1, v.y, v.z)))
					return;
				next = new Vertex(v.x, v.y, v.z-1);
				break;
			case yzFace:
				if(borderContainsEdge(new Vertex(v.x, v.y, v.z), new Vertex(v.x, v.y+1, v.z)))
					return;
				next = new Vertex(v.x, v.y, v.z-1);
				break;
		}
		if(squares.contains(next))
			return;
		
		squares.add(next);
		queue.addLast(next);
	}
	

	private boolean borderContainsEdge(Vertex vertex1, Vertex vertex2) 
	{
		int i = borderVertices.indexOf(vertex1);
		if(i == -1)
			return false;
		if(borderVertices.get((i+1) % borderVertices.size()).equals(vertex2))
			return true;
		if(borderVertices.get((i+borderVertices.size()-1) % borderVertices.size()).equals(vertex2))
			return true;

		return false;
	}

	private void deriveBorderVertices()
	{
		borderVertices.clear();
		Iterator<Vertex> cornerIterator = cornerVertices.iterator();
		Vertex last = null;
		Vertex next = cornerIterator.next();
		while(cornerIterator.hasNext())
		{
			last = next;
			next = cornerIterator.next();
			addPointsBetween(last, next);
		}
		last = next;
		next = cornerVertices.getFirst();
		addPointsBetween(last, next);
		
	}

	private void addPointsBetween(Vertex v1, Vertex v2) 
	{
		if(v1.y == v2.y && v1.z == v2.z) // x changes
		{
			
			for(int i = v1.x; i != v2.x; i=(v1.x < v2.x ? i+1 : i-1))
				borderVertices.add(new Vertex(i, v1.y, v1.z));
		}
		else if(v1.x == v2.x && v1.z == v2.z) // y changes
		{
			for(int i = v1.y; i != v2.y; i=(v1.y < v2.y ? i+1 : i-1))
				borderVertices.add(new Vertex(v1.x, i, v1.z));
		}
		else if(v1.x == v2.x && v1.y == v2.y) // z changes 
		{
			for(int i = v1.z; i != v2.z; i=(v1.z < v2.z ? i+1 : i-1))
				borderVertices.add(new Vertex(v1.x, v1.y, i));
		}
	}
}

class Vertex implements Comparable<Vertex>
{
	public int x;
	public int y;
	public int z;

	public Vertex(int x, int y, int z) 
	{
		this.x = x;
		this.y = y;
		this.z = z;
	}

	public Vertex(Vertex other) 
	{
		this.x = other.x;
		this.y = other.y;
		this.z = other.z;
	}

	@Override
	public int compareTo(Vertex o) 
	{
		if(this.x - o.x != 0)
			return this.x - o.x;
		if(this.y - o.y != 0)
			return this.y - o.y;
		return this.z - o.z;
	}
	
	@Override
	public int hashCode() 
	{
		return (this.x << 20) + (this.y << 10) + this.z;
	}
	
	@Override
	public boolean equals(Object other) 
	{
		if(!(other instanceof Vertex))
			return false;
		Vertex vertex2 = (Vertex)other;

		return this.x == vertex2.x && this.y == vertex2.y && this.z == vertex2.z;
	}
	
	@Override
	public String toString() 
	{
		return "(" + x + "," + y + "," + z + ")";
	}
}
